### ISSUE TITLE
Handle reading, grouping, and aggregating tens of millions of raster summary records

### ISSUE THREAD

**Issue by robyngit** [PermafrostDiscoveryGateway/viz-raster#2]
This issue outlines a *potential* problem that could arise once we start running the rasterization workflow on data that spans the entire arctic. Ideally, we should test and fix this problem before we run the whole workflow (especially if it will take many days to complete.)

## Background

For every GeoTIFF tile that is created during rasterization, a record of summary statistics for each band within the GeoTIFF is saved. Here is an example of the data saved for a single band in a GeoTIFF tile created at zoom level 13:

```
stat                                            iwp_count
bounds  [-148.84277343750014, -148.82080078125014, 70....
min                                                   0.0
max                                                   1.0
mean                                             0.000443
median                                                0.0
std                                              0.021031
var                                              0.000442
sum                                                  29.0
path    /home/thiessenbock/PDG-test/parsl-test/20-file...
tile                            Tile(x=1418, y=887, z=13)
z                                                      13
```

Since the `stat` for this band is `iwp_count` (count of the centroids of each Ice Wedge Polygon within each pixel), then the `min` gives the least number of polygons that any pixel contains, the max gives the highest number of polygons any pixel contains, `sum` gives the total number of ice wedge polygons within that tile, etc.

After creating all of the GeoTIFFs, the `RasterTiler` reads in these records and calculates the `min` and `max` across each `z` level (using the `RasterTiler.get_z_ranges` method). The min and max are then mapped to the 'lowest' and 'highest' color in the color palette for that statistic. (Unless these min and/or max values are explicitly set [in the configuration](https://github.com/PermafrostDiscoveryGateway/viz-staging/blob/f3ac86e588f5f0ce1a36d3795f670d338a71422b/pdgstaging/ConfigManager.py#L180-L186), in which case the calculated min/max are ignored.) The web tile images must use a consistent mapping of value to color so that color has a consistent meaning across tiles at the same zoom level (e.g. 'yellow' should mean '1' for every tile at z=13.)

## The potential problem

Currently, these records are stored in a CSV file. Each time a GeoTIFF is created, rows are appended to the CSV. This strategy works well in the tests we've done so far, where we've been creating thousands or tens of thousands of GeoTIFFs. Once we start processing the entire Arctic, this CSV will grow to over 30 million rows.

Previously, we calculated that there are about 12,225,697 tiles covering land in 50-90 N band at z=13. We also create tiles for the z=12 to z=0 range, resulting in another ~ 4,075,239 tiles (each parent z-level has approx. 25% of the number of tiles as the child level.) This means we'll be creating about 16,300,936 GeoTIFFs each with 2 bands, resulting in a CSV that has 32,601,872 rows.

## Question
**Will whichever hardware we run the workflow on be able to handle reading a 32 million row CSV into memory, then grouping and aggregating it with pandas?**

## Possible solutions

Some possible solutions:

1. Read and write to a database (instead of CSV) with an SQL library like `sqlite3` ([serverless and self-contained](https://www.sqlite.org/different.html), [handles locking and concurrency automatically](https://www.sqlite.org/lockingv3.html))
2. Only read in and process chunks of the CSV at once using panda's [chunking](https://pandas.pydata.org/docs/user_guide/scale.html#use-chunking) feature.
3. Use a library like Dask, which has an [API for working with larger than memory datasets in parallel](https://pandas.pydata.org/docs/user_guide/scale.html#use-other-libraries)

## Next steps

- Create a 32 million row raster summary CSV (perhaps by just duplicating the rows from one of the summaries creating during a test run). Check the filesize of the CSV.
- Try running the `RasterTiler.get_z_ranges` method on this data (on a laptop? on the datateam server?). Time how long it takes, and see if any issues arise.

**Comment by mbjones** [comment]
Thanks for the summary @robyngit . Rather than parsing the whole CSV, isn't it feasible to simply seek to the end, and append a new row on the end of the file? Seems like parsing the whole file is not needed for the append that needs to be done.

In addition to the scale problems, do you also potentially have a locking problem where many threads might be trying to update this file at once? As they seem like append-only operations, it seems like pretty tractable, but one does need to to avoid race conditions.

**Comment by robyngit** [comment]
> Rather than parsing the whole CSV, isn't it feasible to simply seek to the end, and append a new row on the end of the file? Seems like parsing the whole file is not needed for the append that needs to be done.

Yep, and this is exactly [what `RasterTiler` does](https://github.com/PermafrostDiscoveryGateway/viz-raster/blob/b2d7bec1587a223d375f5fe26b308602005938de/pdgraster/RasterTiler.py#L601-L618). The problem is not with writing, but when we need to [read the entire file](https://github.com/PermafrostDiscoveryGateway/viz-raster/blob/b2d7bec1587a223d375f5fe26b308602005938de/pdgraster/RasterTiler.py#L480) then [group and aggregate](https://github.com/PermafrostDiscoveryGateway/viz-raster/blob/b2d7bec1587a223d375f5fe26b308602005938de/pdgraster/RasterTiler.py#L421-L422) to get the overall `min` and `max` for each stat and each z-level. This only needs to happen once.
 
> In addition to the scale problems, do you also potentially have a locking problem where many threads might be trying to update this file at once? As they seem like append-only operations, it seems like pretty tractable, but one does need to to avoid race conditions.

This error hasn't come up but I assume it's probably because the operation is so quick that a conflict is very rare. If we continue with using a CSV for these records, we should lock the file while writing like we do now when saving tiles in viz-staging.

**Comment by robyngit** [comment]
(Added a line to the initial comment to clarify that the problem is with the reading & processing, not the writing to CSV)

**Comment by mbjones** [comment]
That lock might end up being a bottleneck to scaling. An alternative is to use the map-reduce strategy where each thread writes its own results out to an independent file with a single row (and has no dependencies on a shared file), and then a different process concatenates all of the incoming individual files into the master file. That process could even be made to keep the running cumulative stats totals as it parses and appends each line, such that the whole 32M row file would never have to be read in its entirety.

**Comment by mbjones** [comment]
Yeah, in memory. The process could checkpoint its results periodically (say, every 5 minutes or something) so that recovery from a failure or restart would be possible.

**Comment by robyngit** [comment]
@alonakos @rushirajnenuji have you encountered any problems with to the raster summary in parallel? Is this issue still relevant?

**Comment by mbjones** [comment]
Another potential pathway on this is to use a fast serialization format that supports queries, such as Parquet, which can then be processed in place with a tool like duckdb.
